source/extensions/clusters/dns_srv/dns_srv_cluster.cc

```diff
@@ -0,0 +1,351 @@
+
+  dns_lookup_family_ = getDnsLookupFamilyFromCluster(cluster);
+
+  load_assignment_.add_endpoints()->add_lb_endpoints()->set_endpoint_name(
+      dns_srv_cluster_.srv_name());
+
+  ENVOY_LOG(debug, "starting async DNS resolution for {}", dns_srv_cluster_.srv_name());
+}
+
+void DnsSrvCluster::startPreInit() {
+  startResolve();
+  if (!wait_for_warm_on_init_) {
+    onPreInitComplete();
+  }
+}
+
+DnsSrvCluster::~DnsSrvCluster() {
+  if (active_dns_query_) {
+    active_dns_query_->cancel(Network::ActiveDnsQuery::CancelReason::QueryAbandoned);
+  }
+}
+
+void DnsSrvCluster::startResolve() {
+  ENVOY_LOG(debug, "starting async DNS resolution for {}", dns_srv_cluster_.srv_name());
+  info_->configUpdateStats().update_attempt_.inc();
+
+  active_resolve_list_.reset(new ResolveList(*this));
+
+  active_dns_query_ = dns_resolver_->resolveSrv(
+      dns_srv_cluster_.srv_name(),
+      [this](Network::DnsResolver::ResolutionStatus status, absl::string_view details,
+             std::list<Network::DnsResponse>&& response) -> void {
+        active_dns_query_ = nullptr;
+
+        ENVOY_LOG(debug, "Got DNS response, details: {}, status: {}", details,
+                  static_cast<int>(status));
+
+        if (status == Network::DnsResolver::ResolutionStatus::Completed) {
+          for (const auto& dns : response) {
+
+            ENVOY_LOG(debug, "SRV: host: {}, port: {}, weight: {}, prio: {}", dns.srv().target_,
+                      dns.srv().port_, dns.srv().weight_, dns.srv().priority_);
+
+            if (auto address = Envoy::Network::Utility::parseInternetAddressNoThrow(
+                    dns.srv().target_, 0, false);
+                address != nullptr) {
+              // SRV record target is an IP address, not a hostname.
+              ResolveTargetPtr target = ResolveTargetPtr(
+                  new ResolveTarget(*active_resolve_list_, dns_resolver_, dns_lookup_family_,
+                                    dns.srv().target_, dns.srv().port_));
+
+              active_resolve_list_->addResolvedTarget(std::move(target), address);
+            } else {
+              active_resolve_list_->addTarget(ResolveTargetPtr(
+                  new ResolveTarget(*active_resolve_list_, dns_resolver_, dns_lookup_family_,
+                                    dns.srv().target_, dns.srv().port_)));
+            }
+          }
+
+          active_resolve_list_->noMoreTargets();
+        } else {
+          info_->configUpdateStats().update_failure_.inc();
+          resolve_timer_->enableTimer(dns_refresh_rate_ms_);
+          onPreInitComplete();
+        }
+      });
+}
+
+void DnsSrvCluster::allTargetsResolved() {
+  ASSERT(active_resolve_list_.get() != nullptr);
+
+  HostVector new_hosts;
+  absl::flat_hash_set<std::string> all_new_hosts;
+
+  const auto& locality_lb_endpoints = load_assignment_.endpoints()[0];
+  const auto& lb_endpoint_ = load_assignment_.endpoints()[0].lb_endpoints()[0];
```

Comment:
This isn't a member, so don't suffix with `_`

```suggestion
  const auto& lb_endpoint = load_assignment_.endpoints()[0].lb_endpoints()[0];
```